Update foreach binary ops with a scalar list

izdeby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52450 wip
* #49039 [wip] Replace optimizers in torch.optim with the ones from torch.optim._multi_tensor
* #52449 Add type promotion to a fast path for foreach APIs
* #52448 Support broadcasting via slow path
* #48224 [wip] Pushed unsupported scenarios to slow path
* #51170 Updated foreach pointwise ops to support complex and bloat16 on CUDA
* #49714 Update foreach max/min
* #51060 Refactor foreach pointwise ops tests to use OpInfo
* #51059 Refactor foreach binary ops tests with tensor lists to use OpInfo
* #51058 Refactor foreach binary ops tests with scalars to use OpInfo
* #49712 Refactor foreach unary ops tests to use OpInfo
* #49250 Update foreach binary ops with a single scalar and list of tensors
* **#49249 Update foreach binary ops with a scalar list**
* #51131 Refactor ForeachUtils.h

Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
Update logic for a set of foreach APIs with a scalar list
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests

Commit:
Update Binary Ops with scalar lists

[ghstack-poisoned]

Commit:
Update on "Update foreach binary ops with scalar lists"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests
- Update ForeachUtils.h with division check.


[ghstack-poisoned]

Commit:
Update on "Update foreach binary ops with scalar lists"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests
- Update ForeachUtils.h with division check.


[ghstack-poisoned]

Commit:
Update on "Update foreach binary ops with scalar lists"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests
- Update ForeachUtils.h with division check.


[ghstack-poisoned]

Commit:
Update on "Update foreach binary ops with scalar lists"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests
- Update ForeachUtils.h with division check.


[ghstack-poisoned]

Commit:
Update on "Update ForeachBinaryOpScalarList.cu "


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests
- Update ForeachUtils.h with division check.


[ghstack-poisoned]

Commit:
Update on "Update ForeachBinaryOpScalarList.cu "


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests
- Update ForeachUtils.h with division check.


[ghstack-poisoned]

Commit:
Update on "Update ForeachBinaryOpScalarList.cu "


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests
- Update ForeachUtils.h with division check.


[ghstack-poisoned]

Commit:
Update on "Update ForeachBinaryOpScalarList.cu "


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests
- Update ForeachUtils.h with division check.


[ghstack-poisoned]

Commit:
Update on "Update ForeachBinaryOpScalarList.cu "


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests
- Update ForeachUtils.h with division check.


[ghstack-poisoned]

Commit:
Update on "Update ForeachBinaryOpScalarList.cu "


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests
- Update ForeachUtils.h with division check.


[ghstack-poisoned]

Commit:
Update on "Update ForeachBinaryOpScalarList.cu "


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests
- Update ForeachUtils.h with division check.


[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/ForeachUtils.h:
```diff
@@ -151,6 +163,13 @@ bool can_use_fast_route(TensorList tensors1, TensorList tensors2) {
     if (!has_same_attributes(expected_device, {tensors1[i], tensors2[i]})) {
       return false;
     }
+
+    // In case of division, integer inputs will result in float
+    if (division_op) {
+      if (at::isIntegralType(tensors1[i].scalar_type(), /*includeBool*/ true)) {
+        return false;
+      }
+    }
```
It bothers me a little that we have so many `can_use_fast_route` overloads that do roughly the same thing. I wonder if there's a way to refactor the code so that we have even fewer of them

Also, does `bool can_use_fast_route(TensorList tensors1, TensorList tensors2, Scalar scalar)` also need to be changed?

izdeby replied:
refactored. please, let me know if i should break `can_use_fast_route` refactoring into a separate PR.

zou3519 replied:
Nah, this is fine

zou3519 commented on aten/src/ATen/native/ForeachUtils.h:
```diff
@@ -88,7 +88,13 @@ bool has_same_attributes(Device expected_device, TensorList tensors) {
 }
 
-bool will_promote_tensor(const Tensor& tensor, Scalar scalar) {
+bool will_promote_tensor(const Tensor& tensor, Scalar scalar, bool division_op = false) {
```
Division is not the only op that has this problem. IIRC copysign also does int->float promotion. We should probably name the flag something to be more in line with what TensorIterator calls it: https://github.com/pytorch/pytorch/blob/ccd646696bad6a967b56c5097b59ed3da63ee28b/aten/src/ATen/TensorIterator.cpp#L843

zou3519 commented on test/test_foreach.py:
```diff
@@ -667,80 +632,27 @@ def test_bool_scalarlist(self, device, dtype):
-
-                                foreach_bin_op_(tensors, scalars)
-                                self.assertEqual(res, tensors)
+                if dtype in torch.testing.integral_types_and(torch.bool) and foreach_bin_op == foreach_bin_op == torch._foreach_div:
```
`foreach_bin_op == foreach_bin_op == torch._foreach_div`?

izdeby replied:
fixed

zou3519 reviewed (commented):
Questions about behavior

izdeby commented on aten/src/ATen/native/ForeachUtils.h:
```diff
@@ -88,7 +88,13 @@ bool has_same_attributes(Device expected_device, TensorList tensors) {
 }
 
-bool will_promote_tensor(const Tensor& tensor, Scalar scalar) {
+bool will_promote_tensor(const Tensor& tensor, Scalar scalar, bool division_op = false) {
```
100% agree. For H1, one of the first things i will be working on, is implementing real type promotion in MTA core. So all these checks will be going away.

izdeby commented on aten/src/ATen/native/ForeachUtils.h:
```diff
@@ -88,7 +88,13 @@ bool has_same_attributes(Device expected_device, TensorList tensors) {
+bool will_promote_tensor(const Tensor& tensor, Scalar scalar, bool division_op = false) {
+  // In case of division, integer inputs will result in float
+  if (division_op) {
+    if (at::isIntegralType(tensor.scalar_type(), /*includeBool*/ true)) {
+      return true;
+    }
```
import torch

a = torch.tensor([1], dtype=torch.long)
b = torch.tensor([1], dtype=torch.int32)
scalar = 2

print(torch.div(a, 2).dtype) --> torch.float32
print(torch.div(b, 2).dtype) --> torch.float32

izdeby replied:
You can take a look at `test_int_scalarlist` (sorry, not sure how to add a proper link to this). it tests ints inputs as well

zou3519 replied:
Oh I see, you're right, that is my bad

Commit:
Update on "Update ForeachBinaryOpScalarList.cu "


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests
- Update ForeachUtils.h with division check.


[ghstack-poisoned]

Commit:
Update on "Update ForeachBinaryOpScalarList.cu and refactor ForeachUtils.h"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests
- Update ForeachUtils.h with division check and refactored `can_use_fast_route` methods.


[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/cuda/ForeachBinaryOpScalarList.cu:
```diff
@@ -58,16 +58,69 @@ void foreach_tensor_##NAME##_scalarlist_kernel_cuda_(TensorList tensors, at::Arr
+void foreach_tensor_sub_scalarlist_kernel_cuda_(TensorList tensors, at::ArrayRef<Scalar> scalars) {
+    check_foreach_api_restrictions(tensors, scalars);
+
+    TORCH_CHECK(tensors[0].scalar_type() != kBool || !scalars[0].isBoolean(),
+              "Subtraction, the `-` operator, with two bool tensors is not supported. "
+              "Use the `^` or `logical_xor()` operator instead.")
+    TORCH_CHECK(tensors[0].scalar_type() != kBool && !scalars[0].isBoolean(),
+              "Subtraction, the `-` operator, with a bool tensor is not supported. "
+              "If you are trying to invert a mask, use the `~` or `logical_not()` operator instead.");
```
The error messages are more maintainable if we toss them into the same check. It would be nice to unify these error messages with those in BinaryOps.h as well

zou3519 reviewed (commented):
Testing and the bool logic lgtm. We should try to avoid dynamic allocations in `can_use_fast_path` as much as possible though

ngimel commented on aten/src/ATen/native/ForeachUtils.h:
```diff
@@ -88,61 +88,46 @@ bool has_same_attributes(Device expected_device, TensorList tensors) {
-    if (!has_same_attributes(expected_device, {t})) {
-      return false;
+bool will_promote_tensor(const Tensor& tensor, Scalar scalar, bool promote_integer_inputs_to_float = false) {
+  // In case of division, integer inputs will result in float
```
a fair number of unary operations now supports integer to float promotion, grep for `promotes_integers_to_float` in common_methods_invocations.py, which means that you'd need to disable them for a fast path. In general, keeping track of this is pretty hard, so to make it more automatic perhaps testing `for_each` ops should also become OpInfo based? At least supported or unsupported dtypes, we can discuss in more details.

izdeby replied:
Added a few PRs to this stack that will move tests to OpInfo

izdeby committed:
Update on "[wip] Update ForeachBinaryOpScalarList.cu and refactor ForeachUtils.h"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests
- Update ForeachUtils.h with division check and refactored `can_use_fast_route` methods.


[ghstack-poisoned]

Commit:
Update on "[wip] Update ForeachBinaryOpScalarList.cu and refactor ForeachUtils.h"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests
- Update ForeachUtils.h with division check and refactored `can_use_fast_route` methods.


[ghstack-poisoned]

Commit:
Update on "[wip] Update ForeachBinaryOpScalarList.cu and refactor ForeachUtils.h"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests
- Update ForeachUtils.h with division check and refactored `can_use_fast_route` methods.


[ghstack-poisoned]

Commit:
Update on "[wip] Update ForeachBinaryOpScalarList.cu and refactor ForeachUtils.h"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests
- Update ForeachUtils.h with division check and refactored `can_use_fast_route` methods.


[ghstack-poisoned]

Commit:
Update on "[wip] Update ForeachBinaryOpScalarList.cu and refactor ForeachUtils.h"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests
- Update ForeachUtils.h with division check and refactored `can_use_fast_route` methods.


[ghstack-poisoned]

Commit:
Update on "Update foreach binary ops with a scalar list"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests


[ghstack-poisoned]

Commit:
Update on "Update foreach binary ops with a scalar list"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests


[ghstack-poisoned]

Commit:
Update on "Update foreach binary ops with a scalar list"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests


[ghstack-poisoned]

Commit:
Update on "Update foreach binary ops with a scalar list"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests


[ghstack-poisoned]

Commit:
Update on "Update foreach binary ops with a scalar list"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests


[ghstack-poisoned]

Commit:
Update on "Update foreach binary ops with a scalar list"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests


[ghstack-poisoned]

Commit:
Update on "Update foreach binary ops with a scalar list"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests


[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/cuda/ForeachBinaryOpScalarList.cu:
```diff
@@ -66,8 +67,49 @@ std::vector<Tensor> foreach_tensor_##NAME##_scalarlist_kernel_cuda(TensorList te
+
+std::vector<Tensor> foreach_tensor_sub_scalarlist_kernel_cuda(TensorList tensors, at::ArrayRef<Scalar> scalars) {
+    check_foreach_api_restrictions(tensors, scalars);
+    sub_check(tensors[0], scalars[0]);
```
It's not guaranteed that all of the scalars have the same dtype, right? Can a user do the following?

```
tensors = [torch.ones([], dtype=torch.bool) for _ in range(3)]
scalars = [1, 1, True]
_foreach_sub(tensors, scalars)
```

izdeby replied:
Scalars of different dtype are allowed but sub will throw an error in case of boolean scalar

zou3519 commented on test/test_foreach.py:
```diff
@@ -659,80 +650,36 @@ def test_bool_scalarlist(self, device, dtype):
-                            else:
-                                foreach_bin_op_(tensors, scalars)
-                                self.assertEqual(res, tensors)
-                else:
-                    # we dont support complex types on CUDA for now
-                    if (dtype in torch.testing.get_all_complex_dtypes()) and self.device_type == 'cuda':
-                        with self.assertRaisesRegex(RuntimeError, "not implemented for"):
-                            foreach_bin_op_(tensors, scalars)
+                    with self.assertRaisesRegex(RuntimeError, "not implemented for"):
+                        foreach_bin_op(tensors, scalars)
+                    return
```
Why did this have to change? This PR doesn't change the behavior with respect to complex dtype, right? Is this just a simplification?

izdeby replied:
just a simplification

izdeby replied:
I've added support for bool in ForeachBinaryOpScalarList.cu

zou3519 commented on test/test_foreach.py:
```diff
@@ -659,80 +650,36 @@ def test_bool_scalarlist(self, device, dtype):
-                            foreach_bin_op_(tensors, scalars)
+                    with self.assertRaisesRegex(RuntimeError, "Subtraction, the `-` operator"):
+                        foreach_bin_op(tensors, scalars)
 
-                        with self.assertRaisesRegex(RuntimeError, "Subtraction, the `-` operator, with a bool tensor"):
-                            foreach_bin_op(tensors, scalars)
-                    else:
-                        if self.device_type == "cpu":
-                            expected = [torch_bin_op(t, s) for t, s in zip(tensors, scalars)]
-                            res = foreach_bin_op(tensors, scalars)
+                    with self.assertRaisesRegex(RuntimeError, "Subtraction, the `-` operator"):
+                        foreach_bin_op_(tensors, scalars)
+                    continue
```
Should test a scalar list with mixed booleans and integers. E.g. [1, True]. It probably shouldn't go into this test since it's pretty hard to follow already.

izdeby replied:
Added a test

zou3519 commented:
I haven't read through #49250 (the next PR in this stack) but I wanted to ask first: What is the difference between #49250 and this PR? Both of the descriptions look the same:

This PR
```
Update binary ops to support boolean type
Update sub ops to throw an error in case of bool subtraction
Update division op to check for type promotion.
Update tests
```

#49250
```
Update foreach_sub error checking
Update division logic to check for type promotion
Updated tests
```

izdeby committed:
Update on "Update foreach binary ops with a scalar list"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests


[ghstack-poisoned]

Commit:
Update on "Update foreach binary ops with a scalar list"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests


[ghstack-poisoned]

izdeby commented:
> I haven't read through #49250 (the next PR in this stack) but I wanted to ask first: What is the difference between #49250 and this PR? Both of the descriptions look the same:
> 
> This PR
> 
> ```
> Update binary ops to support boolean type
> Update sub ops to throw an error in case of bool subtraction
> Update division op to check for type promotion.
> Update tests
> ```
> 
> #49250
> 
> ```
> Update foreach_sub error checking
> Update division logic to check for type promotion
> Updated tests
> ```

In this PR, we are updating the sub/div logic for APIs that work with scalar lists and in the next one APIs with a scalar and/or tensor list. Updating names and description for both PRs

Commit:
Update on "Update foreach binary ops with a scalar list"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
Update logic for 2 set of foreach APIs with a scalar list
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests


[ghstack-poisoned]

Commit:
Update on "Update foreach binary ops with a scalar list"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
Update logic for 2 set of foreach APIs with a scalar list
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests


[ghstack-poisoned]

Commit:
Update on "Update foreach binary ops with a scalar list"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
Update logic for a set of foreach APIs with a scalar list
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests


[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/cuda/ForeachBinaryOpScalarList.cu:
```diff
@@ -67,8 +68,53 @@ std::vector<Tensor> foreach_tensor_##NAME##_scalarlist_kernel_cuda(TensorList te
 FOREACH_BINARY_OP_SCALARLIST(mul, std::multiplies);
-FOREACH_BINARY_OP_SCALARLIST(div, std::divides);
+
+// In the case of division, integer inputs will result in float. 
+// Currently multi tensor apply can only return result of the same type as input.
```
nit: It would be good to mention "This does not use FOREACH_BINARY_OP_SCALARLIST because <everything else in the comment>". It's easy to make the inference but just to clarify

izdeby replied:
Done

zou3519 commented on aten/src/ATen/native/cuda/ForeachBinaryOpScalarList.cu:
```diff
@@ -67,8 +68,53 @@ std::vector<Tensor> foreach_tensor_##NAME##_scalarlist_kernel_cuda(TensorList te
+
+// In the case of division, integer inputs will result in float. 
+// Currently multi tensor apply can only return result of the same type as input.
+void foreach_tensor_div_scalarlist_kernel_cuda_(TensorList tensors, at::ArrayRef<Scalar> scalars) {
```
Also nit: is it possible to update FOREACH_BINARY_OP_SCALARLIST to take three arguments, where the last argument is div_op? Then we would be able to use FOREACH_BINARY_OP_SCALARLIST with division instead of rewriting the division logic

izdeby replied:
Done

zou3519 commented on aten/src/ATen/native/cuda/ForeachBinaryOpScalarList.cu:
```diff
@@ -67,8 +68,53 @@ std::vector<Tensor> foreach_tensor_##NAME##_scalarlist_kernel_cuda(TensorList te
+    return foreach_binary_op<std::divides>(tensors, scalars);
+}
+
+// In the case of subtraction, we dont allow scalar to be boolean following the torch.sub logic
```
It would be good to mention "This does not use FOREACH_BINARY_OP_SCALARLIST because ". It's easy to make the inference but just to clarify that this is a special case

izdeby replied:
Done

zou3519 commented on test/test_foreach.py:
```diff
@@ -664,80 +655,64 @@ def test_bool_scalarlist(self, device, dtype):
-                                foreach_bin_op_(tensors, scalars)
+                    with self.assertRaisesRegex(RuntimeError, "not implemented for"):
+                        foreach_bin_op(tensors, scalars)
+                    return
```
Should this be a continue?

izdeby replied:
Done

zou3519 commented on test/test_foreach.py:
```diff
@@ -664,80 +655,64 @@ def test_bool_scalarlist(self, device, dtype):
+        for N in N_values:
+            for foreach_bin_op, foreach_bin_op_, torch_bin_op in self.bin_ops:
+                tensors = self._get_test_data(device, dtype, N)
+                scalars = [True for _ in range(N)]
+                scalars[0] = 1
+                scalars[1] = 1.1
+                scalars[2] = 3 + 5j
```
nit: can be done succinctly in one line: `scalars = [1, 1.1, 3 + 5j] + [True for _ in range(N - 3)]`

izdeby replied:
Done

zou3519 commented on test/test_foreach.py:
```diff
@@ -664,80 +655,64 @@ def test_bool_scalarlist(self, device, dtype):
-
-                                foreach_bin_op_(tensors, scalars)
-                                self.assertEqual(res, tensors)
+                expected = [torch_bin_op(t, s) for t, s in zip(tensors, scalars)]
+                res = foreach_bin_op(tensors, scalars)
+                self.assertEqual(expected, res)
```
This only tests correctness for the out-of-place variant, right? Should it also test correctness for the in-place variant?

izdeby replied:
Done

zou3519 reviewed (commented):
This looks pretty reasonable, it's great we are deleting all the special cases in the testing logic. I had some comments around the tests

Commit:
Update on "Update foreach binary ops with a scalar list"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
Update logic for a set of foreach APIs with a scalar list
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests


[ghstack-poisoned]

Commit:
Update on "Update foreach binary ops with a scalar list"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
Update logic for a set of foreach APIs with a scalar list
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests


[ghstack-poisoned]

Commit:
Update on "Update foreach binary ops with a scalar list"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
Update logic for a set of foreach APIs with a scalar list
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests


[ghstack-poisoned]

Commit:
Update on "Update foreach binary ops with a scalar list"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
Update logic for a set of foreach APIs with a scalar list
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests


[ghstack-poisoned]

Commit:
Update on "Update foreach binary ops with a scalar list"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
Update logic for a set of foreach APIs with a scalar list
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests


[ghstack-poisoned]

izdeby committed:
Update on "Update foreach binary ops with a scalar list"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
Update logic for a set of foreach APIs with a scalar list
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests


[ghstack-poisoned]

zou3519 commented on test/test_foreach.py:
```diff
@@ -671,80 +662,72 @@ def test_bool_scalarlist(self, device, dtype):
+                else:
+                    with self.assertRaisesRegex(RuntimeError, "can't be cast to the desired output type"):
+                        foreach_bin_op_(tensors, scalars)
+                    continue
```
Is the continue necessary here?

izdeby committed:
Update on "Update foreach binary ops with a scalar list"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
Update logic for a set of foreach APIs with a scalar list
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests


[ghstack-poisoned]

izdeby committed:
Update on "Update foreach binary ops with a scalar list"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
Update logic for a set of foreach APIs with a scalar list
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests


[ghstack-poisoned]

izdeby committed:
Update on "Update foreach binary ops with a scalar list"


Differential Revision: [D25502939](https://our.internmc.facebook.com/intern/diff/D25502939)

--------
Update logic for a set of foreach APIs with a scalar list
- Update binary ops to support boolean type
- Update sub ops to throw an error in case of bool subtraction
- Update division op to check for type promotion.
- Update tests


[ghstack-poisoned]
